hep/proposals/0000-decorators.md

```diff
@@ -0,0 +1,299 @@
+    # using the rshift operator if they wish to define dependencies amongst
+    # tasks that don't share variables directly
+    return WorkerOutput(value=final_task.result)
+```
+
+
+## Steps Template
+
+```python
+from hera.workflows import WorkflowTemplate
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class CalculatorInput(hio.Input):
+    x: int
+    y: int
+    operation: Literal["add", "sub"] = "add"
+
+@wt.script  # Adds a new script template to the workflow template called calculator
+def calculator(calc_input: CalculatorInput) -> hio.Output:
+    if calc_input.operation == "add":
+        return hio.Output(calc_input.x + calc_input.y)
+    return hio.Output(calc_input.x - calc_input.y)
+    
+
+# This defines the another template input
+class FiboInput(hio.Input):
+    num: int
+
+class FiboOutput(hio.Output):
+    num: int
+
+@wt.entrypoint
+@wt.steps  # Adds a new script template to the workflow template called fibonacci
+def fibonacci(fibo: FiboInput) -> FiboOutput:
+    # we will need to validate that users can only run/define certain kinds of expressions here
+    # and throw warnings if they are using syntax that is invalid
+    valid_number = hio.expr(fibo.num > 1)
+    previous_num = calculator(CalculatorInput(x=fibo.num, y=1, operation="sub"), when=valid_number)
+    previous_fibo = fibonacci(FiboInput(num=previous_num.result), when=valid_number)
+    current = calculator(CalculatorInput(x=fibo.num, y=previous_fibo.num), when=valid_number)
+    return FiboOutput(num=valid_number.check(current.num, 1))
+```
+
+
+
+## Container Template
+
+```python
+from hera.workflows import WorkflowTemplate, Container
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class MyInput(hio.Input):
+    user: str = "Hera"
+
+class MyOutput(hio.Output):
+    container_greeting: Annotated[str, Parameter(name="container-greeting")]
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"], args=["echo {{input.parameters.user}}"])
+def basic_hello_world(my_input: MyInput) -> hio.Output:
+    ...
+
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"])
+def advanced_hello_world(my_input: MyInput, template: Container) -> MyOutput:
+    # hio.output serves two functions. One, it allows users to create the output class
+    # which allows output path variables to be referenced.
+    # Second, it can be used to "mock" a container template when running locally and allow us to
+    # inject outputs based on inputs.
+    output: MyOutput = hio.output(my_input)
+    # template is a special variable that allows you to reference the template itself and modify
+    # its attributes. This is especially useful when trying to reference input params in args
+    template.args = [f"echo {hio.name(my_input.user)} > {hio.path(output.container_greeting)}"]
```

Comment:
I think then you could literally run the command through Python? Maybe we should expose that somehow - as in, if there's no implementation when running locally, just do

```py
subprocess.run(template.args)
```

Also might be good to point out for the reader that mocking would be good in the case of external calls to avoid maintaining two sets of code.